### DISCUSSION TITLE
Building Grid children at runtime

### DISCUSSION THREAD

**Discussion by GingerAvalanche** [AvaloniaUI/Avalonia#8449] [Q&A] [answered]
I'm making a localization editor that needs one set of controls per language to be localized, and I don't know how many languages the user will have until runtime. Each set of controls is a TextBlock, TextBox, and Button.

I'm trying to get implement this using a grid - 3 columns, and a row for each language. Since I don't know how many languages they have until they've picked a folder to load from, I'm just using an empty Grid and then calling an Update function on my UserControl view once I have the info, which populates the grid's children.

```
        public void Update(string[] langs)
        {
            RootGrid.Children.Clear();

            RowDefinition rowDef = new() { Height = GridLength.Star };
            Binding zero = new("0");
            Binding one = new("1");
            Binding two = new("2");
            Binding saveLoc = new("SaveLoc");

            for (int i = 0; i < langs.Length; i++)
            {
                Binding row = new($"{i}");
                RootGrid.RowDefinitions.Add(rowDef);
                TextBlock textBlock = new();
                textBlock[!Grid.RowProperty] = row;
                textBlock[!Grid.ColumnProperty] = zero;
                textBlock[!TextBlock.TextProperty] = new Binding($"Languages[{i}]");
                RootGrid.Children.Add(textBlock);
                RootGrid.Children.Add(new TextBox()
                {
                    [!Grid.RowProperty] = row,
                    [!Grid.ColumnProperty] = one,
                    [!TextBox.TextProperty] = new Binding($"LocText{i}"),
                });
                RootGrid.Children.Add(new Button()
                {
                    [!Grid.RowProperty] = row,
                    [!Grid.ColumnProperty] = two,
                    [!Button.CommandProperty] = saveLoc,
                    [!Button.CommandParameterProperty] = new Binding(i.ToString()),
                    Content = "Apply",
                });
            }
        }
```

The line `textBlock[!Grid.RowProperty] = row;` throws an ExpressionParseException, unexpected end of expression. I'm not sure what I'm doing wrong, though, as the expression expects an IBinding, which is what I'm giving it. Am I using the `[!Grid.RowProperty]` wrong, or is this a bug?

**Comment by maxkatz6** [comment]
This binding `Binding row = new($"{i}");` can be read as `Binding row = new("1");`
Which is wrong binding path expression.
If you want indexer path, you need `new("[1]")` or `new($"[{i}]")`;

**Reply by GingerAvalanche** [reply]
Oh. I didn't even realize it *was* an indexer. In XAML, it's just a string. This doesn't seem to do what I want, though.

Basically, I'm looking for the code-behind equivalent of `<TextBlock Grid.Column="0" Grid.Row="0" Text="{Binding Languages[0]}"/>` where each 0 can vary.

When setting the bindings as indexers, like you suggested, it creates this
![image](https://user-images.githubusercontent.com/32864080/177434166-52913857-b824-40d3-aaf7-86cd10ff43ea.png)
but it should create this 
![image](https://user-images.githubusercontent.com/32864080/177434197-67de3a5e-350d-4d2c-9fb9-e820bbdec95b.png)

I'm pretty sure that means it's not setting the Grid.Column and Grid.Row properly.

Note that this is the Grid definition in XAML, which is where the referenced ColumnDefinitions are.
```
    <Grid Name="RootGrid" Margin="100">
      <Grid.ColumnDefinitions>
        <ColumnDefinition Width="1*"/>
        <ColumnDefinition Width="5*"/>
        <ColumnDefinition Width="1*"/>
      </Grid.ColumnDefinitions>
    </Grid>
```

Also, separate issue I'm now having: How to set `[!Button.CommandParameterProperty]` to something other than a string. Basically, to recreate `<Button.CommandParameter><sys:Int32>1</sys:Int32></Button.CommandParameter>` in code-behind.

**Comment by GingerAvalanche** [comment, marked as answer]
The indexer binding was on the right track, it was just referencing the wrong thing. Unfortunately, I wasn't able to make it reference the *right* thing with the Binding constructor, so I had to get creative.

I ended up creating the indexer bindings for the grid rows and columns like so...
```
            IndexerBinding zero = new(RootGrid, new AttachedProperty<int>("Column", typeof(Grid), new(0)), BindingMode.Default);
            IndexerBinding one = new(RootGrid, new AttachedProperty<int>("Column", typeof(Grid), new(1)), BindingMode.Default);
            IndexerBinding two = new(RootGrid, new AttachedProperty<int>("Column", typeof(Grid), new(2)), BindingMode.Default);
```
and the indexer binding for the Button's CommandProperty like so...
```
                ((Button)RootGrid.Children[2])[!Button.CommandParameterProperty] = new IndexerBinding(RootGrid.Children[2], new StyledProperty<int>("CommandParameter", typeof(Button), new(i)), BindingMode.Default);
```

These are pretty horrifying to read, though, so if anyone has a better way of doing it, I'd appreciate the feedback.

**Reply by maxkatz6** [reply]
You don't need and shouldn't create a new attached/styled property. Reuse existing instances.

**Reply by rosseg** [reply]
IndexerBinding is no longer accessible due to its protection level. Any idea how to set the grid rows and columns at runtime since `Binding row = new("1");` doesn't work? 
I get the same result as the OP:
https://github.com/AvaloniaUI/Avalonia/discussions/8449#discussioncomment-3088387
It appears the default row/column index of 0 is kept.
